### PR TITLE
Standardize error handling for functions

### PR DESCRIPTION
**Pull request by kbnopro**

### CODE REVIEW

**Comment by kbnopro** [comment]
Also moved components to app to improve DX.

File: src/middleware.ts [hunk 1]
```diff
@@ -10,9 +10,8 @@ export const middleware = async (request: NextRequest) => {
   const unauthenticatedOnlyPrefixes = ["/login"] as const;
 
   if (adminOnlyPrefixes.some((path) => pathname.startsWith(path))) {
-    const token = request.cookies.get("token");
-    const userInfo = await getUser(token?.value);
-    if (!userInfo.ok || userInfo.user.status !== "Admin") {
+    const userInfo = await verifyCodefunWithMemo();
```
Comment (DuCanhGH) [inline src/middleware.ts:13]: Why? Is `cookies()` intended for use in middleware?

File: src/utils/handleCatch.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      ok: false,
+      message: e.message,
+      status: statusCode,
+    };
+  } else {
+    console.error("Unexpected error: ", e);
+    if (e instanceof Error) {
+      return {
+        ok: false,
+        message: e.message,
```
Comment (DuCanhGH) [inline src/utils/handleCatch.ts:18]: Can this return an internal error?

File: src/features/problems/queries/getProblem.ts [hunk 3]
```diff
@@ -53,14 +52,13 @@ export const getProblem = async (
...
-    } else {
-      console.error(e);
-    }
-    throw new Error("Internal Server Error");
+    return handleCatch(e);
   }
 };
 
-export const getMemoProblem = cache(getProblem);
+export const getProblemWithMemo = cache<typeof getProblem>(getProblem);
```
Comment (DuCanhGH) [inline src/features/problems/queries/getProblem.ts:64]: Why do we need to specify the generic here? Shouldn't it be inferred?

File: src/features/auth/actions/actionLogin.ts
Comment (DuCanhGH) [inline src/features/auth/actions/actionLogin.ts:1]: Why was this file removed?

File: src/features/auth/actions/actionAuthenticate.ts
Comment (DuCanhGH) [inline src/features/auth/actions/actionAuthenticate.ts:1]: Why was this file removed?

File: src/features/auth/actions/actionLogout.ts
Comment (DuCanhGH) [inline src/features/auth/actions/actionLogout.ts:1]: Why was this file removed?

File: src/features/auth/api/getUser.ts
Comment (DuCanhGH) [inline src/features/auth/api/getUser.ts:1]: Why was this file removed?

File: src/features/auth/actions/index.ts
Comment (DuCanhGH) [inline src/features/auth/actions/index.ts:1]: Why was this file removed?

File: src/components/NavigationBar/UserInfo/client.tsx [hunk 4]
```diff
@@ -24,8 +25,29 @@ export const UserInfoClient = ({
...
+    if (shouldDisplayError) {
+      const timeout = setTimeout(() => {
+        setShouldDisplayError(false);
+      }, 5000);
+      return () => {
+        clearTimeout(timeout);
+      };
+    }
+  }, [shouldDisplayError, setShouldDisplayError]);
+
```
Comment (DuCanhGH) [inline src/components/NavigationBar/UserInfo/client.tsx:50]: Can these two effects... be combined? This looks like an unnecessary extra render to me.

File: src/app/problems/create/CreateProblemForm/Inputs.tsx [hunk 5]
```diff
@@ -5,38 +5,54 @@ import { useActionState, useEffect, useState } from "react";
...
+      const timer = setTimeout(() => {
+        setShouldDisplaySuccess(false);
       }, 5000);
-      return () => {
-        clearTimeout(timeout);
-      };
+      return () => clearTimeout(timer);
     }
-  }, [displayState]);
+  }, [shouldDisplaySuccess]);
```
Comment (DuCanhGH) [inline src/app/problems/create/CreateProblemForm/Inputs.tsx:55]: Can these four effects be rewritten into two pair of effects?

File: src/app/login/LoginForm/Inputs.tsx [hunk 6]
```diff
@@ -4,41 +4,33 @@ import { useActionState, useEffect, useState } from "react";
...
-        setDisplayState(initialState);
-      }, 5000);
-      return () => {
-        clearTimeout(timeout);
-      };
+    if (!state.ok && !!state.message) {
+      setShouldDisplayMessage(true);
     }
-  }, [displayState]);
+  }, [state]);
```
Comment (DuCanhGH) [inline src/app/login/LoginForm/Inputs.tsx:32]: Can these two effects be combined?

**Review by DuCanhGH — CHANGES_REQUESTED** [review]

File: src/middleware.ts (hunk 1 above)
Comment (kbnopro) [reply src/middleware.ts:13]: oops, my bad

File: src/features/auth/actions/actionLogin.ts
Comment (kbnopro) [reply src/features/auth/actions/actionLogin.ts:1]: BY MISTAKES 😭

File: src/features/auth/actions/actionAuthenticate.ts
Comment (kbnopro) [reply src/features/auth/actions/actionAuthenticate.ts:1]: Because we no longer need actionAuthenticate. This is intended but the rest is not.

File: src/utils/handleCatch.ts (hunk 2 above)
Comment (kbnopro) [reply src/utils/handleCatch.ts:18]: Yeah make more sense

File: src/features/problems/queries/getProblem.ts (hunk 3 above)
Comment (kbnopro) [reply src/features/problems/queries/getProblem.ts]: Agree

File: src/components/NavigationBar/UserInfo/client.tsx (hunk 4 above)
Comment (kbnopro) [reply src/components/NavigationBar/UserInfo/client.tsx]: Will try to next time 😄
